### ISSUE TITLE
Allow changing window title bar color in macOS 

### ISSUE THREAD

**Issue by dominiklohmann** [jwilm/alacritty#686]
Recent iTerm2 versions allow selecting a color for the title bar on macOS to make it blend in with the terminal background. It doesn't hide the title bar, but makes it look a lot better. 

Is there any chance of getting this for alacritty? Or, can someone point me in the right direction towards implementing this myself? 

![lmbchxs](https://user-images.githubusercontent.com/4488655/28492796-2c6a02e6-6f0a-11e7-917e-aec4cd3fc4a3.png)

This can be achieved by using the following flags on NSWindow, assuming the window background color is set accordingly (example in Swift). 

```swift
window.titlebarAppearsTransparent = true
window.titleVisibility = .hidden
window.styleMask.insert(NSWindowStyleMask.fullSizeContentView)
```

**Comment by monouser7dig** [comment]
this actually looks really nice

**Comment by asilvadesigns** [comment]
yes please

**Comment by tbodt** [comment]
This would be best implemented in glutin.

**Comment by metalelf0** [comment]
It would look great! It would be great to allow theme authors to specify a `title_bar_color` under `primary` namespace, and fallback to `primary.background` otherwise.

**Comment by dominiklohmann** [comment]
The positive feedback definitely shows that aesthetics are highly valued.

However, looking through glutin I have no idea where to start, so unless someone else has a good idea on where to apply the necessary changes this is probably put on hold.

**Comment by mcginty** [comment]
Yay aesthetics!

At first glance, it looks like glutin uses a `winit` library (same author), which is used for cross-platform window initialization, but `winit` doesn't seem to have a lot of support for providing custom NSWindow flags in macOS.

You can [remove all window decorations](https://github.com/tomaka/winit/blob/master/src/platform/macos/window.rs#L339) by adding a

```rust
.with_decorations(false)
```

into https://github.com/jwilm/alacritty/blob/master/src/window.rs#L188, but I don't immediately see a way outside of submitting a PR to `winit` to do what is needed, except possibly by directly accessing the cocoa `NSWindow` object directory through [`os::macos::WindowExt`](https://github.com/tomaka/winit/blob/master/src/os/macos.rs#L9).

**Comment by mcginty** [comment]
Decided to play with it a bit.

![screen shot 2017-07-27 at 8 41 13 pm](https://user-images.githubusercontent.com/373823/28670587-f9350230-730b-11e7-8915-2fa4ea392b53.png)

I got it partially working by modifying `winit` in a horribly hackish way, but even with that work would still need to be done to add padding to the top, and change the title font color.

**Comment by sodiumjoe** [comment]
FYI, I made [an issue](https://github.com/tomaka/winit/issues/323) on winit.

**Comment by sodiumjoe** [comment]
FYI, looks like https://github.com/tomaka/winit/pull/389 will make this possible. Some more discussion here: https://github.com/jwilm/alacritty/pull/947#issuecomment-360457297

**Comment by sodiumjoe** [comment]
FYI, I opened a PR on winit to allow resizability on `with_decorations(false)` windows, but there was an objection as to whether that should be the default behavior of `with_decorations(false)`, so I limited the scope to just adding a subclass that makes it possible: https://github.com/tomaka/winit/pull/408

Once that's merged, there will be an option to pull in `cocoa` and manipulate the window masks directly by calling `ns::get_window()`. That will allow creation of resizable windows that look like this:

<img width="605" alt="screenshot 2018-03-16 20 52 24" src="https://user-images.githubusercontent.com/845857/37551391-f4c59226-295b-11e8-8626-970fb0c089eb.png">

Notice there are no borders, no titlebar, square corners, and no drop shadow.

There's also this open PR https://github.com/tomaka/winit/pull/389 which will allow windows that look like this:

![screenshot 2018-03-16 20 54 51](https://user-images.githubusercontent.com/845857/37551427-7bc6b44e-295c-11e8-9b38-cffdd33ede0f.png)

This has a transparent titlebar with a "full size content window" meaning the titlebar will have the same color background as the content, and retains the titlebar buttons, rounded corners, and drop shadow.

I also have a PR into the above PR https://github.com/robertgzr/winit/pull/1 to allow the same as above, but also hide the titlebar buttons, to allow a window that looks like this:

![screenshot 2018-03-16 20 59 53](https://user-images.githubusercontent.com/845857/37551453-004eacbc-295d-11e8-9f87-13085d4a2155.png)

This basically looks like a window with no titlebar, but retains rounded corners and drop shadow.

I'm trying to get those PRs moving. I'm not sure how we would expose those options from alacritty, though. Should we expose all of these options? Or alias `decorations: false` to one of these? (Current `decorations: false` doesn't allow any resizability, so imo is not very useful.)

**Comment by jedahan** [comment]
Providing a middle ground between configurability and ease of use - if we use the term decorations to cover borders, titlebar, corner, and shadow, I'd propose from first screenshot to last:
```
decorations: none
decorations: transparent
decorations: buttonless
```

'none' covers the fullscreen or quake-like terminal use case.
'transparent' covers floating 
'buttonless' covers minimal floating

There are two use cases that I think we do not have to cover (until someone pipes up)
- buttonless without transparency
- no decorations + not resizeable (what we have now with decorations: false)

also `decorations: default` would just be regular old window like we have now

**Comment by sodiumjoe** [comment]
quick update, I opened a new PR so the relevant PRs that this depends on are:
https://github.com/tomaka/winit/pull/408
https://github.com/tomaka/winit/pull/423

**Comment by sodiumjoe** [comment]
@jedahan I think that's a really good summary. I agree with the options you've presented, but I'm not sure how this would work for other platforms. I'm also not sure if `decorations` correctly describes what this does anymore. I wonder if it would make more sense to present a new field in the config that's macos specific. Something for @jwilm and @chrisduerr to weight in on, probably.

**Comment by sodiumjoe** [comment]
https://github.com/tomaka/winit/pull/423 is merged and https://github.com/tomaka/winit/pull/408 is approved, hopefully merged very soon. I'd like to start work on a PR into alacritty to integrate those changes. @jwilm does [jedehan's suggestion](https://github.com/jwilm/alacritty/issues/686#issuecomment-373933767) sound good to you for macos-specific config options?

**Comment by chrisduerr** [comment]
If a macos-specific setting is added it should probably be similar to `thin_strokes` where it is present in both configs but only does something in one.

This should help us with the long-term plan to merge the two default configurations together once the font configuration RFC is implemented.

**Comment by sodiumjoe** [comment]
while waiting for https://github.com/tomaka/winit/pull/408 to merge, I started an implementation of the above: https://github.com/sodiumjoe/alacritty/commit/1d20fae37f2adc9102452fc08b467ed242da0471

Any feedback would be welcome!

**Comment by sodiumjoe** [comment]
https://github.com/tomaka/winit/pull/408 has merged, we're not just waiting for a release

**Comment by wavded** [comment]
Just a heads up, looks like winit released these changes in `0.12` a few days ago: https://github.com/tomaka/winit/blob/master/CHANGELOG.md.  And BTW, thanks for working on this @sodiumjoe !

**Comment by sodiumjoe** [comment]
oh nice, ok, i'll get a PR up to integrate that change into alacritty

**Comment by chrisduerr** [comment]
We might have to wait for glutin because we're not using winit directly, but if that's updated too it should be all good to go!

Thanks for putting in all that nice work on macos @sodiumjoe :)

**Comment by sodiumjoe** [comment]
Yeah, it looks like glutin is still on winit@0.11. There's a PR up for updating it to @0.12 tho: https://github.com/tomaka/glutin/pull/1010

**Comment by sodiumjoe** [comment]
Ok, glutin updated and published, I've made a PR for this: #1241

**Comment by ChadTaljaardt** [comment]
Thank you @sodiumjoe

**Comment by monouser7dig** [comment]
Thanks sodiumjoe! 
> as a follow up it would be nice if it was possible to not only make the titlebar transparent but also make the text start below the titelbar ❤️

**Comment by sodiumjoe** [comment]
@monouser7dig I think that should be achieved by granular padding config for `top,right,bottom,left` instead of just `x,y`. Either way it should probably be a separate issue.

**Comment by pascalw** [comment]
Indeed top-only padding would be very welcome when #686 is merged and running with `decorations: transparent`. Currently I need `padding.y: 50` to make sure the terminal contents don't run under the window buttons, but this has the side effect of not using the bottom 50 pixels off the screen either.

**Comment by jedahan** [comment]
@yech1990 that would be cool, wanna open another github issue to request it, and pointers to example applications that behave like you'd want?

**Comment by elithrar** [comment]
VSCode's "custom" title bars on both macOS, Windows & Linux should be good inspiration: https://code.visualstudio.com/updates/v1_28#_custom-title-and-menu-bar-on-windows-by-default
